public/app/features/dashboard-scene/scene/PanelRepeaterGridItem.tsx

```diff
@@ -0,0 +1,248 @@
+import { css } from '@emotion/css';
+import React, { useMemo } from 'react';
+
+import { config } from '@grafana/runtime';
+import {
+  VizPanel,
+  SceneObjectBase,
+  VariableDependencyConfig,
+  SceneVariable,
+  SceneGridLayout,
+  SceneVariableSet,
+  ConstantVariable,
+  SceneComponentProps,
+  SceneGridItemStateLike,
+  SceneGridItemLike,
+  sceneGraph,
+  MultiValueVariable,
+  VariableValueSingle,
+} from '@grafana/scenes';
+import { GRID_CELL_HEIGHT, GRID_CELL_VMARGIN } from 'app/core/constants';
+
+interface PanelRepeaterGridItemState extends SceneGridItemStateLike {
+  source: VizPanel;
+  repeatedPanels?: VizPanel[];
+  variableName: string;
+  itemHeight?: number;
+  repeatDirection?: RepeatDirection | string;
+  maxPerRow?: number;
+}
+
+export type RepeatDirection = 'v' | 'h';
+
+export class PanelRepeaterGridItem extends SceneObjectBase<PanelRepeaterGridItemState> implements SceneGridItemLike {
+  protected _variableDependency = new VariableDependencyConfig(this, {
+    variableNames: [this.state.variableName],
+    onVariableUpdatesCompleted: this._onVariableChanged.bind(this),
+  });
+
+  private _ignoreNextStateChange = false;
+  private _isWaitingForVariables = false;
+
+  public constructor(state: PanelRepeaterGridItemState) {
+    super(state);
+
+    this.addActivationHandler(() => this._activationHandler());
+  }
+
+  private _activationHandler() {
+    this._subs.add(this.subscribeToState((newState, prevState) => this._handleGridResize(newState, prevState)));
+
+    // If we our variable is ready we can process repeats on activation
+    if (sceneGraph.hasVariableDependencyInLoadingState(this)) {
+      this._isWaitingForVariables = true;
+    } else {
+      this._performRepeat();
+    }
+  }
+
+  private _onVariableChanged(changedVariables: Set<SceneVariable>, dependencyChanged: boolean): void {
+    if (dependencyChanged) {
+      this._performRepeat();
+      return;
+    }
+
+    // If we are waiting for variables and the variable is no longer loading then we are ready to repeat as well
+    if (this._isWaitingForVariables && !sceneGraph.hasVariableDependencyInLoadingState(this)) {
+      this._isWaitingForVariables = false;
+      this._performRepeat();
+    }
+  }
+
+  /**
+   * Uses the current repeat item count to calculate the user intended desired itemHeight
+   */
+  private _handleGridResize(newState: PanelRepeaterGridItemState, prevState: PanelRepeaterGridItemState) {
+    if (this._ignoreNextStateChange) {
+      this._ignoreNextStateChange = false;
+      return;
+    }
```

Comment:
Can you elaborate on this? Feels like handling some issue you have encountered that's not explicit in this PR.

Reply:
@dprokop well the object subscribes to state changes to detect grid changes by the user (resize). But it also changes its own height when number of repeated panels change, so this was to ignore that (and prevent infinite recursion). 

But now that you brought this up I found a more elegant solution.

If I add this to the state change handler 
```ts
 if (stateChange.itemHeight !== this.state.itemHeight) {
    this.setState(stateChange);
 }
```

I avoid the recursion and can remove the _ignoreNextStateChange field.